consumer/consumererror/README.md

```diff
@@ -0,0 +1,231 @@
+# Consumer errors
+
+This package contains error types that should be returned by a consumer when an
+error occurs while processing telemetry. The error types included in this
+package provide functionality for communicating details about
+the error for use upstream in the pipeline. Ideally the error returned by a
+component in its `consume` function should be from this package.
+
+## Error classes
+
+**Retryable**: Errors are retryable if re-submitting data to a sink may result
+in a successful submission.
+
+**Permanent**: Errors are permanent if submission will always fail for the
+current data. Errors are considered permanent unless they are explicitly marked
+as retryable.
+
+## Use cases
+
+**Retry logic**: Errors should be allowed to include information necessary to
+perform retries.
+
+**Indicating partial success**: Errors can indicate that not all items were
+accepted, for example as in an OTLP partial success message.
+
+**Communicating network error codes**: Errors should allow embedding information
+necessary for the Collector to act as a proxy for a backend, i.e. relay a status
+code returned from a backend in a response to a system upstream from the
+Collector.
+
+## Current targets for using errors
+
+**Receivers**: Receivers should be able to consume multiple errors downstream
+and determine the best course of action based on the user's configuration. This
+may entail either keeping the retry queue inside of the Collector by having the
+receiver keep track of retries, or may involve having the caller manage the
+retry queue by returning a retryable network code with relevant information.
+
+**exporterhelper**: When an exporter returns a retryable error, the
+exporterhelper can use this information to manage the sending queue if it is
+enabled. Permanent errors will be forwarded back up the pipeline.
+
+## Creating Errors
+
+Errors can be created by calling `consumererror.New(err, opts...)` where `err`
+is the underlying error, and `opts` is one of the provided options for supplying
+additional metadata:
+
+- `consumererror.WithRetry`
+- `consumererror.WithPartial`
```

Comment:
I'll add a section about this. I think we have to assume that the user knows what they are doing, even if we likely consider it an error. The rule is that an error is retryable if it includes retry information, so in this case we have to assume that a component had partial success, but also supports retries on partial success, even if OTLP considers partial success permanent.

Reply:
Let's assume is retryable, what data are you retrying? Based on my understanding "partial" only contains the number of failed entries, so you cannot retry because you don't know what failed.

Reply:
You would have to retry the whole payload, and the backend would need to support filtering out the items that have already been accepted. From the perspective of a component upstream from where this error was created, we don't know anything about where the data will be sent and will just have to work with the available context.

I don't expect anyone to actually do this, but I also don't see any meaningful way for us to prevent it. The best we could likely do would be to swallow one of the options and log a warning on the console if we can determine a precedence for the errors, but I would still prefer to leave that to upstream components to deal with. I'd also prefer to keep things resilient and not return an error while creating the error.

Even if we break out the error cases into separate types, if you have to join anything with `errors.Join`, you're back in the same place where contradictory errors can be placed next to each other.